pennylane/devices/default_qubit.py

```diff
@@ -142,6 +179,142 @@ def _apply_operation(self, operation):
+        """
+        return self._apply_phase(state, axes, -1)
+
+    def _apply_hadamard(self, state, axes, **kwargs):
+        """Apply the Hadamard gate by combining the results of applying the PauliX and PauliZ gates.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        state_x = self._apply_x(state, axes)
+        state_z = self._apply_z(state, axes)
+        return SQRT2INV * (state_x + state_z)
+
+    def _apply_s(self, state, axes, inverse=False):
+        return self._apply_phase(state, axes, 1j, inverse)
+
+    def _apply_t(self, state, axes, inverse=False):
+        return self._apply_phase(state, axes, TPHASE, inverse)
+
+    def _apply_cnot(self, state, axes, **kwargs):
+        """Applies a CNOT gate by slicing along the first axis specified in ``axes`` and then
+        applying an X transformation along the second axis.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
+
+        state_x = self._apply_x(state[sl_1], axes=target_axes)
+        return self._stack([state[sl_0], state_x], axis=axes[0])
+
+    def _apply_swap(self, state, axes, **kwargs):
+        """Applies a SWAP gate by performing a partial transposition along the specified axes.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        all_axes = list(range(len(state.shape)))
+        all_axes[axes[0]] = axes[1]
+        all_axes[axes[1]] = axes[0]
+        return self._transpose(state, all_axes)
+
+    def _apply_cz(self, state, axes, **kwargs):
+        """Applies a CZ gate by slicing along the first axis specified in ``axes`` and then
+        applying a Z transformation along the second axis.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
+
+        state_z = self._apply_z(state[sl_1], axes=target_axes)
+        return self._stack([state[sl_0], state_z], axis=axes[0])
+
+    def _apply_phase(self, state, axes, parameters, inverse=False):
```

Comment:
Could the `PhaseShift` operation be made to use this method, rather than `apply_diagonal_unitary`?